### ISSUE TITLE
Inheritance 2

### ISSUE THREAD

**Issue by Udayv56** [Udayv56/CPP#24]
Single inheritance is a feature of object-oriented programming where a derived class can inherit the members (attributes and methods) of a single base class. The derived class can extend or modify the functionality of the base class while inheriting its existing features. In single inheritance, there is a one-to-one relationship between the base class and the derived class.

**CODE-**
#include<iostream>
using namespace std;

class Base{
    int data1; // not inheritable
    public:
        int data2;
        void setData();
        int getData1();
        int getData2();
};

void Base :: setData(void){
    data1 = 10;
    data2 = 20;
    
}

int Base :: getData1(){
    return data1;
}

int Base :: getData2(){
    return data2;
}

class Derived : public Base{ // if inheritance will be done privately then setData() couldnt be alled directly
    int data3;
    public:
        void process();
        void display();
};

void Derived :: process(){
    data3 = getData1() + data2; // data1 is private to Base class
}

void Derived :: display(){
    cout<<data3<<endl;
}
    

int main(){
    
    Derived d;
    d.setData();
    d.process();
    d.display();
    return 0;
}

**OUTPUT-**
30

Now, let's go through the provided code step by step:

1. The code starts by including the necessary header files, including `<iostream>` for input/output operations.

2. The `Base` class is defined, which has two private data members: `data1` (which is not inheritable) and `data2` (which is inheritable). It also has three member functions: `setData()`, `getData1()`, and `getData2()`. 

3. The `setData()` function is defined outside the class. It sets the values of `data1` and `data2` to 10 and 20, respectively.

4. The `getData1()` function is defined outside the class. It returns the value of `data1`.

5. The `getData2()` function is defined outside the class. It returns the value of `data2`.

6. The `Derived` class is defined, which is derived from the `Base` class using the public access specifier. This means that the public and protected members of the `Base` class will be accessible in the `Derived` class.

7. The `Derived` class has one private data member `data3` and two member functions: `process()` and `display()`.

8. The `process()` function is defined outside the class. It calculates the value of `data3` by accessing the private member `data1` from the base class using the `getData1()` function, and adds it to the value of `data2` from the base class.

9. The `display()` function is defined outside the class. It simply prints the value of `data3`.

10. In the `main()` function, an object `d` of the `Derived` class is created.

11. The `setData()` function is called on the object `d`, which sets the values of `data1` and `data2` in the base class.

12. The `process()` function is called on the object `d`, which calculates the value of `data3` using the values from the base class.

13. The `display()` function is called on the object `d`, which prints the value of `data3` to the console.

14. The program ends with a `return 0` statement.

In summary, this code demonstrates single inheritance where the `Derived` class inherits the members of the `Base` class. It shows how the derived class can access the public members of the base class, even though the base class has private members that are not directly accessible in the derived class.
